src/modules/slack/slack.c

```diff
@@ -0,0 +1,298 @@
+ *
+ * Kamailio is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "slack.h"
+
+MODULE_VERSION
+
+static char *_slmsg_buf = NULL;
+static int mod_init(void);
+static void mod_destroy(void);
+
+static int buf_size = 4096;
+static char *slack_url = NULL;
+static char *slack_channel = SLACK_DEFAULT_CHANNEL;
+static char *slack_username = SLACK_DEFAULT_USERNAME;
+static char *slack_icon = SLACK_DEFAULT_ICON;
+
+/**
+ * Exported functions
+ */
+static cmd_export_t cmds[] = {
+	{"slack_send",	  		(cmd_function)slack_send1,			1, slack_fixup,  0, ANY_ROUTE},
+	{0, 0, 0, 0, 0, 0}
+};
+
+
+/**
+ * Exported parameters
+ */
+static param_export_t mod_params[] = {
+	{ "slack_url", 			PARAM_STRING|USE_FUNC_PARAM, (void*)_slack_url_param},
+	{ "channel",			PARAM_STRING, &slack_channel }, // channel starts with #
+	{ "username",			PARAM_STRING, &slack_username },
+	{ "icon_emogi",			PARAM_STRING, &slack_icon },
+	{0, 0, 0}
+};
+
+/**
+ * Module description
+ */
+struct module_exports exports = {
+	"slack",        /* 1 module name */
+	DEFAULT_DLFLAGS, /* 2 dlopen flags */
+	cmds,            /* 3 exported functions */
+	mod_params,      /* 4 exported parameters */
+	0,               /* 5 exported RPC functions */
+	0,         		 /* 6 exported pseudo-variables */
+	0,               /* 7 response function */
+	mod_init,        /* 8 module initialization function */
+	0,      		 /* 9 per-child init function */
+	mod_destroy      /* 0 destroy function */
+};
+
+/**
+ * Module init
+ */
+static int mod_init(void) {
+	LM_INFO("slack module init\n");
+
+	_slmsg_buf = (char*)pkg_malloc((buf_size+1)*sizeof(char));
+	if(_slmsg_buf==NULL)
+	{
+		PKG_MEM_ERROR;
+		return -1;
+	}
+	return(0);
+}
+
+/**
+ * Module destroy
+ */
+static void mod_destroy() {
+	LM_INFO("slack module destroy\n");
+	if(_slmsg_buf)
+		pkg_free(_slmsg_buf);
+	if(slack_url)
+		pkg_free(slack_url);
+	return;
+}
+
+/**
+ * send message with curl
+ * @return 0 on success, -1 on error
+ */
+static int _curl_send(const char* uri, str *post_data)
+{
+	int datasz;
+	char* send_data;
+	CURL *curl_handle;
+	CURLcode res;
+	// LM_DBG("sending to[%s]\n", uri);
+
+	datasz = snprintf(NULL, 0, BODY_FMT, slack_channel, slack_username, post_data->s, slack_icon);
+	send_data = (char*)pkg_malloc((datasz+1)*sizeof(char));
+	if(send_data==NULL) {
+        LM_ERR("Error: can not allocate pkg memory [%d] bytes\n", datasz);
+        return -1;
+    }
+    snprintf(send_data, datasz+1, BODY_FMT, slack_channel, slack_username, post_data->s, slack_icon);
```

Comment:
see above

Reply:
thank you for your notes. I am wondering, if the above snprintf check is passed, what could be a reason for this one to fail? do we really need to error check this one?

Reply:
This of course depends on the working of snprintf, the docs unfortunately do not give much more details. But generally it is good "defensive" programming practice to check for errors on outside library calls that you do not control/own.